src/model64.c

```diff
@@ -10,88 +11,376 @@
+            primitive->indices = PTR_ENCODE(model->meshes, primitive->indices);
         }
-        model->meshes[i].primitives = PTR_ENCODE(model, model->meshes[i].primitives);
+        model->meshes[i].primitives = PTR_ENCODE(model->meshes, model->meshes[i].primitives);
     }
+    model->nodes = PTR_ENCODE(model, model->nodes);
     model->meshes = PTR_ENCODE(model, model->meshes);
-}
-
-void model64_free(model64_t *model)
-{
-    model64_unload(model);
-    if (model->magic == MODEL64_MAGIC_OWNED) {
+    if(model->skins)
+    {
+        for(uint32_t i=0; i<model->num_skins; i++)
+        {
+            model->skins[i].joints = PTR_ENCODE(model->skins, model->skins[i].joints);
+        }
+        model->skins = PTR_ENCODE(model, model->skins);
+    }
+    if(model->magic == MODEL64_MAGIC_OWNED) {
         #ifndef NDEBUG
-        // To help debugging, zero the model structure
-        memset(model, 0, sizeof(model64_t));
+        // To help debugging, zero the model data structure
+        memset(model, 0, sizeof(model64_data_t));
         #endif
 
         free(model);
     }
 }
 
+void model64_free(model64_t *model)
+{
+    if(--model->data->ref_count == 0)
+    {
+        unload_model_data(model->data);
+    }
+    free(model);
+}
+
 uint32_t model64_get_mesh_count(model64_t *model)
 {
-    return model->num_meshes;
+    return model->data->num_meshes;
 }
+
 mesh_t *model64_get_mesh(model64_t *model, uint32_t mesh_index)
 {
-    return &model->meshes[mesh_index];
+    return &model->data->meshes[mesh_index];
+}
+
+uint32_t model64_get_node_count(model64_t *model)
+{
+    return model->data->num_nodes;
+}
+
+model64_node_t *model64_get_node(model64_t *model, uint32_t node_index)
+{
+    return &model->data->nodes[node_index];
+}
+
+model64_node_t *model64_search_node(model64_t *model, const char *name)
+{
+    for(uint32_t i=0; i<model->data->num_nodes; i++)
+    {
+        char *node_name = model->data->nodes[i].name;
+        if(node_name && !strcmp(node_name, name))
+        {
+            return &model->data->nodes[i];
+        }
+    }
+    return NULL;
+}
+
+static uint32_t get_node_idx(model64_t *model, model64_node_t *node)
+{
+    return node-model->data->nodes;
+}
+
+void model64_set_node_pos(model64_t *model, model64_node_t *node, float x, float y, float z)
+{
+    uint32_t node_idx = get_node_idx(model, node);
+    assertf(node_idx < model->data->num_nodes, "Setting position of invalid node.");
+    float *dst = model->transforms[node_idx].transform.pos;
+    dst[0] = x;
+    dst[1] = y;
+    dst[2] = z;
+    calc_node_local_matrix(model, node_idx);
+    update_node_matrices(model);
+}
+
+void model64_set_node_rot(model64_t *model, model64_node_t *node, float x, float y, float z)
+{
+    float cr = cosf(x * 0.5);
+    float sr = sinf(x * 0.5);
+    float cp = cosf(y * 0.5);
+    float sp = sinf(y * 0.5);
+    float cy = cosf(z * 0.5);
+    float sy = sinf(z * 0.5);
```

Comment:
please use `fm_sincosf` here

Reply:
I don't have fm_sincosf in my local branch of libdragon.